Automatically create tick formatters for str and callable inputs.

toddrjen opened the pull request:
## PR Summary

This allows `matplotlib.axis.Axis.set_major_formatter` and `matplotlib.axis.Axis.set_minor_formatter` to accept a ``str`` or function (really any ``callable``) in addition to a `matplotlib.ticker.Formatter`.  This will create and use a `matplotlib.ticker.StrMethodFormatter` and `matplotlib.ticker.FuncFormatter`, respectively.  This significantly reduces the barrier to using custom formatters in simple cases.

This was previously [discussed](https://mail.python.org/pipermail/matplotlib-devel/2020-March/001326.html) on the matplotlib-devel mailing list.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

tacaswell commented:
Thanks @toddrjen !

To try and make sure feature PRs don't fall through the cracks, we starting to assign champions who are responsible for making sure the review process goes through, @story645 has agreed to champion this PR.

toddrjen commented:
@tacaswell @story645 Great, thank you!

I don't know why the test coverage has dropped, I added new tests that, as far as I can tell, cover the new code paths.

Assuming this is a good approach, the big outstanding issue in my mind is whether this should be used more in the documentation.  There are several examples using the old-style `FormatStrFormatter` that could be switched to using this approach.

story645 commented:
@toddrjen can you please provide some examples here cause I'm struggling a bit in finding 'em in the PR

anntzer commented on lib/matplotlib/axis.py:
```diff
@@ -1529,37 +1529,68 @@ def set_major_formatter(self, formatter):
         """
         Set the formatter of the major ticker.
 
+        In addition to a `~matplotlib.ticker.Formatter` instance,
```
As argued in the mailing list (not repro'ing the discussion here) I'd like to argue in favor of
```
This accepts either a `.Formatter` instance, or a callable, which is
automatically wrapped in a `.FuncFormatter`.  To use a format string
(e.g. "{x:.4f}"), pass the corresponding bound string method
(`"{x:.4f}".format`).
```
which I don't actually think is harder to understand (in one case the user needs to learn about StrMethodFormatter, which is a highly matplotlib specific concept; in the other case, about bound methods, which is also somewhat advanced, but at least much more widely applicable).

(Then you can remove all the mentions of StrMethodFormatter.)

toddrjen replied:
In my opinion the `StrMethodFormatter` is the most basic formatter.  There is probably a reason the documentation routinely uses the `FormatStrFormatter` rather than the `FuncFormatter`.  Further, in the documentation the string-related formatter are usually described first.  Learning to work with strings is one of the first thing new python developers encounter, while passing around function objects is much more advanced.

So although `StrMethodFormatter` is a new function, it isn't a new concept.  Once a beginning user sees they can use strings, they won't have to worry about functions in most cases.  Using a bound method like this, on the other hand, is a new concept for many users, one they wouldn't necessarily have to deal with otherwise.  This goes against the purpose of this change, which is to make things simpler and easier to use.

That is similar to the reason I intentionally don't even mention callables.  A beginning users isn't going to know what that is and is just going to get confused, while a more advanced user is going to understand that callables can be used wherever functions are normally used.

So although this makes the code simpler, it only really saves two lines.  And I think this comes at the expense of making things more difficult for beginning users, exactly the sort of people I was trying to help with this change.

jklymak replied:
I agree with @toddrjen here.  We want to make this less arcane, not more arcane, and telling folks to pass the `format` method of a string is pretty arcane.   Most people want to just do: `format_my_ticks('%01.3f')`.  

WRT this, you apparently need the `x` in `'{x}'`?   This also needs to be documented?  If you are going to just refer people to the underlying Formatter classes to understand the syntax needed, then you lose a lot of the simplification here.

anntzer replied:
I don't feel strongly enough to block over this, but enough to not approve :shrug:

toddrjen commented:
@story645  One example for the `str` approach used in the documentation is:

```Python
axs[5].xaxis.set_major_formatter('{x} km')
```

Which would replace:

```Python
from matplotlib import ticker
axs[5].xaxis.set_major_formatter(ticker.StrMethodFormatter('{x} km'))
```

One for the function approach used in the documentation:

```Python
axs[3].xaxis.set_major_formatter(lambda x, pos: str(x-5))
```

Which would replace:

```Python
from matplotlib import ticker
axs[3].xaxis.set_major_formatter(ticker.FuncFormatter(lambda x, pos: str(x-5)))
```

jklymak commented:
In the pr it’s not clear that you need a pos argument to your function or why.

toddrjen commented:
@jklymak Where, exactly, do you think this should be mentioned?  Can you comment on the specific lines or blocks where you think this should be changed?  In most cases I am just using or following up on the existing description for the `FuncFormatter`.

jklymak commented:
> @jklymak Where, exactly, do you think this should be mentioned? Can you comment on the specific lines or blocks where you think this should be changed? In most cases I am just using or following up on the existing description for the `FuncFormatter`.

It needs to be documented somewhere, otherwise how are people to know what a valid callable is?

toddrjen commented:
@jklymak Please see the updated version.  I have added more explicit documentation.

toddrjen commented:
I have added a new commit where I modified a bunch of existing examples and tutorials, and the previous documentation, to make better use of the new syntax.  These changes may be more controversial so I added them to a separate commit so they would be easier to roll back.  

Please take a look and see what you think.

I don't know why the tests are failing.  It doesn't appear to be related to my commit but I am not sure.

toddrjen commented:
@jklymak All tests are passing now.  Is there anything else I need to do or is this ready?

toddrjen commented:
@story645  All tests are passing now. Is there anything else I need to do or is this ready?

story645 commented:
Sorry I owe still you feedback on this, just wanted to note that I'm +1 on this idea since learning that spottily is basically using this sort of thing as their [motivating example](https://labs.spotify.com/2018/11/15/introducing-chartify-easier-chart-creation-in-python-for-data-scientists/) for writing a  new charting library.

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+        this also accepts a ``str`` or function.
+
+        For a ``str`` a `~matplotlib.ticker.StrMethodFormatter` is used.
+        The field used for the value must be labeled ``'x'`` and the field used
```
I understand you're using field to be more general, but I'd go a drop more technical here 'cause I think it can be clearer:
``` 
the tick value argument must be named ``'x'`` and the tick position argument must be labeled ``'pos'``
```

toddrjen replied:
"Field" is the term used by the upstream Python documentation, so I think we should stick with that.  It really isn't an "argument".

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
```
For a function, a

toddrjen replied:
Okay

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
+        The function should take in two inputs (a tick value ``x`` and a
```
must, not should? also two arguments (or two fields if you use the terminology mentioned above - I guess above all keep to consistent terminology)

toddrjen replied:
Fixed

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+        In addition to a `~matplotlib.ticker.Formatter` instance,
+        this also accepts a ``str`` or function.
+
+        For a ``str`` a `~matplotlib.ticker.StrMethodFormatter` is used.
+        The field used for the value must be labeled ``'x'`` and the field used
+        for the position must be labeled ``'pos'``.
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
+        The function should take in two inputs (a tick value ``x`` and a
+        position ``pos``), and return a string containing the corresponding
+        tick label.
```
I think this level of detail should only exist in one place, especially since you cross link to the canonical source.

toddrjen replied:
I was specifically told by someone else that I needed to add more detail here.

story645 replied:
Ok, so we discussed this on the call and consensus is that the canonical explanation should be in `set_major_formatter`  and the other places should cross link to it. `_set_formatter` doesn't need the full documentation since it is a private method. The cost of having to maintain 3 copies of the same docs outweighs the potential clarity gains since the docs can be cross reference.

toddrjen replied:
Okay, I did this.  Although it won't help people as much when using `help` directly.

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
         """
-        cbook._check_isinstance(mticker.Formatter, formatter=formatter)
+        Set the formatter of the major or minor ticker.
+
+        In addition to a `~matplotlib.ticker.Formatter` instance,
+        this also accepts a ``str`` or function.
+
+        For a ``str`` a `~matplotlib.ticker.StrMethodFormatter` is used.
+        The field used for the value must be labeled ``'x'`` and the field used
+        for the position must be labeled ``'pos'``.
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
+        The function should take in two inputs (a tick value ``x`` and a
+        position ``pos``), and return a string containing the corresponding
+        tick label.
+        See the  `~matplotlib.ticker.FuncFormatter` documentation for
+        more information.
```
Same as above, it should be in one canonical place with cross references. Maybe a short hand in comments for devs.

toddrjen replied:
Again, I was told I add more detail here.

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+            cbook._check_isinstance(mticker.Formatter, formatter=formatter)
+
         if (isinstance(formatter, mticker.FixedFormatter)
                 and len(formatter.seq) > 0
```
this s may have been in the code before, but why does it having a length matter for this check?

toddrjen replied:
I don't really understand FixedFormatter and wouldn't want to touch it.

story645 commented on lib/matplotlib/axis.py:
```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
-        self.minor.formatter = formatter
+
+        if level == self.major:
+            self.isDefault_majfmt = False
```
really unclear what this attributes mean/do so not sure if the code is doing the right thing here

toddrjen replied:
This attribute is being used by the original, individual functions.  I just combined them.  But I added a test to make sure they are being set properly.

story645 commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -204,6 +204,28 @@ def test_formatter_ticker():
     ax.autoscale_view()
 
 
+def test_auto_formatter():
```
can you break this into two tests, one for the string functionality and one for the function functionality?

toddrjen replied:
Done

toddrjen commented on doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst:
```diff
@@ -0,0 +1,7 @@
+Allow tick formatters to be set with str or function inputs
```
They can already be set with strings and functions, just indirectly.  I want to make it clear that this is directly as inputs.  And this is an optional approach, which isn't clear from your version.  So I don't think your version is as clear about what is actually changing here.

toddrjen commented on doc/users/next_whats_new/2020-03-06-auto-tick-formatters.rst:
```diff
@@ -0,0 +1,7 @@
+now accept `str` or function inputs in addition to `~.ticker.Formatter`
+instances. For a `str` a `~.ticker.StrMethodFormatter` is automatically
+created and used. For a function a `~.ticker.FuncFormatter` is automatically
+created and used.
```
Fixed.

toddrjen commented on examples/mplot3d/surface3d.py:
```diff
@@ -31,9 +31,27 @@
+matplotlib.pyplot.subplots
+matplotlib.axis.Axis.set_major_formatter
+matplotlib.axis.Axis.set_major_locator
+matplotlib.ticker.StrMethodFormatter
```
Added.

toddrjen commented on examples/pyplots/dollar_ticks.py:
```diff
@@ -7,16 +7,15 @@
+# Use automatic StrMethodFormatter
+ax.yaxis.set_major_formatter('${x:1.2f}')
 
 for tick in ax.yaxis.get_major_ticks():
```
Okay, I changed it.

toddrjen commented on examples/showcase/anatomy.py:
```diff
@@ -24,13 +24,14 @@
 ax.yaxis.set_minor_locator(AutoMinorLocator(4))
-ax.xaxis.set_minor_formatter(FuncFormatter(minor_tick))
+# FuncFormatter is created and used automatically
+ax.xaxis.set_minor_formatter(minor_tick)
```
The "new machinery" is solely in the `set_minor_formatter` and `set_major_formatter` methods.  The existing `FuncFormatter` behavior has not been changed.  Everything that works before will still work.

toddrjen commented on tutorials/introductory/lifecycle.py:
```diff
@@ -187,9 +187,14 @@
-# functions by using the :class:`ticker.FuncFormatter` class. Below we'll
-# define a function that takes an integer as input, and returns a string
-# as an output.
+# functions. Below we'll define a function that takes an integer as input, and
```
Okay I changed this.

toddrjen commented on lib/matplotlib/ticker.py:
```diff
@@ -415,8 +421,8 @@ class StrMethodFormatter(Formatter):
 
-    The field used for the value must be labeled *x* and the field used
-    for the position must be labeled *pos*.
+    The field used for the tick value must be labeled *x* and the field used
```
Again, this is the term used by the upstream Python documentation.  It isn't really an argument, so I wouldn't want to use that term.

toddrjen commented on lib/matplotlib/ticker.py:
```diff
@@ -159,6 +159,12 @@
   ax.yaxis.set_major_formatter(ymajor_formatter)
   ax.yaxis.set_minor_formatter(yminor_formatter)
 
+In addition to `.Formatter`, `~.Axis.set_major_formatter` and
```
Done.

toddrjen commented on tutorials/text/text_intro.py:
```diff
@@ -350,27 +349,29 @@
 #  Finally, we can specify functions for the formatter using
-# `matplotlib.ticker.FuncFormatter`.
+# `matplotlib.ticker.FuncFormatter`.  Further, like
+# `matplotlib.ticker.StrMethodFormatter`, simply passing a function will
```
Done.

toddrjen commented on tutorials/text/text_intro.py:
```diff
@@ -285,16 +285,16 @@
 # format string) or `matplotlib.ticker.FormatStrFormatter` (old-style '%'
-# format string) and passed it to the ``ax.xaxis``.
+# format string) and passed it to the ``ax.xaxis``.  A
+# `matplotlib.ticker.StrMethodFormatter` can also be created by simply
```
Done

toddrjen commented on examples/ticks_and_spines/custom_ticker1.py:
```diff
@@ -21,13 +20,26 @@
-plt.xticks(x, ('Bill', 'Fred', 'Mary', 'Sue'))
+# Use automatic FuncFormatter creation
+ax.yaxis.set_major_formatter(millions)
+ax.bar(x, money)
```
Changed this.

toddrjen commented on examples/ticks_and_spines/tick-formatters.py:
```diff
@@ -34,46 +34,87 @@ def setup(ax, title):
+# and must return a ``str``  This creates a FuncFormatter automatically.
+setup(axs[1], title="lambda x, pos: str(x-5)")
+axs[1].xaxis.set_major_formatter(lambda x, pos: str(x-5))
 
```
Okay.

toddrjen commented on tutorials/intermediate/artists.py:
```diff
@@ -652,16 +652,13 @@ class in the matplotlib API, and the one you will be working with most
-ax.yaxis.set_major_formatter(formatter)
+ax.yaxis.set_major_formatter('${x:1.2f}')
 
 for tick in ax.yaxis.get_major_ticks():
```
I'll need to check on this.

toddrjen replied:
It appears to be possible but I need to do some work to get it working.

toddrjen commented on lib/matplotlib/ticker.py:
```diff
@@ -159,6 +159,12 @@
 
+In addition to `.Formatter`, `~.Axis.set_major_formatter` and
+`~.Axis.set_minor_formatter` also accept a ``str`` or function.  In the case of
+a ``str``, a `.StrMethodFormatter` will be automatically created and applied
```
Okay, done.

toddrjen commented on examples/ticks_and_spines/tick-formatters.py:
```diff
@@ -34,46 +34,87 @@ def setup(ax, title):
-fig, axs = plt.subplots(7, 1, figsize=(8, 6))
+fig, axs = plt.subplots(9, 1, figsize=(8, 8))
 
-# Null formatter
```
Okay, done.

toddrjen commented on tutorials/introductory/lifecycle.py:
```diff
@@ -187,9 +187,14 @@
+#
+# For this function, the ``x`` argument is the original tick label and ``pos``
+# is the tick position.  We will only use ``x`` here but both arguments are
+# needed.
```
There are two ways I can think of to "magic" this away.  First I can try using two arguments, if there is a TypeError I can try one argument, then fail.  Alternatively I can calculate the number of arguments.  The first approach is fast but may silence valid errors.  The second is slower but more reliable.  Which would you prefer?

story645 replied:
what's the time cost on trying to compute the errors?

toddrjen replied:
On the order of tens of microseconds.  Using a try...except is on the order of microseconds.

toddrjen commented:
@story645 All the discussed changes should be implemented.  Please take a look.  I am still waiting on feedback on a few issues.

story645 commented:
Thanks for your patience, been swamped with Passover. @jklymak and @anntzer have your concerns been addressed?

anntzer commented:
As stated above I'm basically fine with the PR, but don't like one very specific point (direct support for strings) so don't particularly want to approve it with that feature.

jklymak commented on .flake8:
```diff
@@ -180,6 +180,7 @@ per-file-ignores =
     examples/misc/svg_filter_line.py: E402
     examples/misc/svg_filter_pie.py: E402
     examples/misc/table_demo.py: E201
+    examples/mplot3d/surface3d.py: E402
```
This looks unrelated to the PR?  Maybe something wrong w a rebase?

toddrjen commented:
@jklymak 
> Do we want to replace _every_ instance of `StrMethodFormatter` and `FuncFormatter` by this idiom? There is some advantage to pointing out the ticker module and the existence of the suite of Formatters.

I didn't replace every instance.  `examples/ticks_and_spines/tick-formatters.py`  still explains how to use the existing method.

story645 commented:
Hi Todd, thanks for hanging in there! We're aiming to get this into 3.3 (rc out in a week or so).  

Giving it some more thought, keeping the current formatter signature seems fine as it's consistent with the docs and what folks will find on stackoverflows. Can revisit in a later PR. 

Do you feel it's ready to merge or do you have any additions to make?

story645 commented:
Also now thinking the arguments fix should actually be on FuncFormatter since it's not clear why it needs two https://matplotlib.org/_modules/matplotlib/ticker.html#FuncFormatter

toddrjen commented:
@story645 If all the tests pass I think it is ready to merge.
